lib/puppet/functions/ensure_packages.rb

```diff
@@ -6,34 +6,56 @@
 # third argument to the ensure_resource() function.
 Puppet::Functions.create_function(:ensure_packages, Puppet::Functions::InternalFunction) do
   # @param packages
-  #   The packages to ensure are installed. If it's a Hash it will be passed to `ensure_resource`
+  #   The packages to ensure are installed.
   # @param default_attributes
   #   Default attributes to be passed to the `ensure_resource()` function
   # @return [Undef] Returns nothing.
   dispatch :ensure_packages do
     scope_param
-    param 'Variant[String[1], Array[String[1]], Hash[String[1], Any]]', :packages
+    param 'Variant[String[1], Array[String[1]]]', :packages
     optional_param 'Hash', :default_attributes
     return_type 'Undef'
   end
 
-  def ensure_packages(scope, packages, default_attributes = nil)
-    if default_attributes
+  # @param packages
+  #   The packages to ensure are installed. The keys are packages and values are the attributes specific to that package.
+  # @param default_attributes
+  #   Default attributes. Package specific attributes from the `packages` parameter will take precedence.
+  # @return [Undef] Returns nothing.
+  dispatch :ensure_packages_hash do
+    scope_param
+    param 'Hash[String[1], Any]', :packages
+    optional_param 'Hash', :default_attributes
+    return_type 'Undef'
+  end
+
+  def ensure_packages(scope, packages, default_attributes = {})
+    Array(packages).each do |package_name|
       defaults = { 'ensure' => 'installed' }.merge(default_attributes)
-      if defaults['ensure'] == 'present'
-        defaults['ensure'] = 'installed'
-      end
-    else
-      defaults = { 'ensure' => 'installed' }
+
+      # `present` and `installed` are aliases for the `ensure` attribute. If `ensure` is set to either of these values replace
+      # with `installed` by default but `present` if this package is already in the catalog with `ensure => present`
+      defaults['ensure'] = default_ensure(package_name) if ['present', 'installed'].include?(defaults['ensure'])
+
+      scope.call_function('ensure_resource', ['package', package_name, defaults])
     end
+    nil
+  end
 
-    if packages.is_a?(Hash)
-      scope.call_function('ensure_resources', ['package', packages.dup, defaults])
-    else
-      Array(packages).each do |package_name|
-        scope.call_function('ensure_resource', ['package', package_name, defaults])
-      end
+  def ensure_packages_hash(scope, packages, default_attributes = {})
+    packages.each do |package, attributes|
+      ensure_packages(scope, package, default_attributes.merge(attributes))
     end
     nil
   end
+
+  private
+
+  def default_ensure(package_name)
+    if call_function('defined_with_params', "Package[#{package_name}]", { 'ensure' => 'present' })
```

Comment:
yeah, I did try and figure out a more generic solution, but also figured the less I touch, the less chance of subtly breaking anything!

This is effectively my 2nd PR on the way to fixing this issue.  The first one was a while back and turned the function into a modern API version.

Whilst I'm not super happy with the hackiness of this solution, as a whole with the refactoring in this PR combined with the work done when [moving from the old API](https://github.com/puppetlabs/puppetlabs-stdlib/commit/82b74e20fb999b6c59669379669eaaf0fc0bba38#diff-59069ddd386b5c5dab21830e69d57f21313760745be35d5662d76b7a576b6862L7-L50) I think the final result is still a lot nicer than what we originally had.

The breaking change in 8.0.0 has been a blocker for several people in upgrading stdlib, so hopefully this is a good enough fix and can be merged.

Reply:
I also ran into it myself, so I think this would be good start.

Reply:
I drilled into this for a bit and got closer, but not completely there. I got stuck trying to figure out how to get a list of known aliases and couldn't get there in this context. What I'd like to do is update this check, https://github.com/puppetlabs/puppetlabs-stdlib/blob/b2ace0d0c3796bab520b7f650b716474cba243d6/lib/puppet/parser/functions/defined_with_params.rb#L65

to do something smarter, like `res[key].aliases.include? value` instead of the straight equality check it does now.

In other contexts, I can get a list of aliases using `Puppet::Parameter::Value.aliases` but I can't quite get there from here yet.

Reply:
@binford2k that's exactly what my line of thought was, except I gave up on figuring out the `.aliases` bit. As I said: those parts look like private APIs.